greenplumpython/func.py

```diff
@@ -170,292 +158,332 @@ def serialize(self) -> str:
                     s = to_pg_const(self._args[i])  # type: ignore
                 args_string_list.append(s)
             args_string = ",".join(args_string_list)
-        return f"{self._func_name}({args_string})"
+        return f"{self._func.qualified_name}({args_string})"
 
     def __call__(self, group_by: Optional[TableRowGroup] = None, table: Optional[Table] = None):
         return ArrayFunctionExpr(
-            self._func_name,
+            self._func,
             self._args,
             group_by=group_by,
-            as_name=self._as_name,
             table=table,
             db=self._db,
-            is_return_comp=self.is_return_comp,  # type: ignore
+            returning_composite=self._returning_composite,
         )
 
 
-def function(name: str) -> Callable[..., FunctionExpr]:
-    """
-    A wrap in order to call function
+# The parent class for all database functions.
+# It is not a Callable by design to prevent misuse.
+class _AbstractFunction:
+    def __init__(
+        self,
+        wrapped_func: Optional[Callable[..., Any]],
+        name: Optional[str],
+        schema: Optional[str],
+    ) -> None:
+        if not (
+            (name is not None and len(name) <= 63)
+            or (wrapped_func is not None and len(wrapped_func.__name__) <= 63)
+        ):
+            raise Exception(
+                "Function name should be no longer than 63 bytes."
+            )  # i.e. NAMEDATALEN - 1 in PostgreSQL
+        qualified_name = (
+            (name if schema is None else f"{schema}.{name}")
+            if name is not None
+            else f"pg_temp.{wrapped_func.__name__}"
+            if wrapped_func is not None
+            else None
+        )
+        assert qualified_name is not None
+        assert (
+            qualified_name not in _global_scope
+        ), f'Function named "{qualified_name}" has been defined before.'
+        self._qualified_name = qualified_name
+        _global_scope[qualified_name] = self
 
-    Example:
-            .. code-block::  Python
+    @property
+    def qualified_name(self) -> str:
+        return self._qualified_name
 
-                generate_series = gp.function("generate_series", db)
+    def _infer_db(self, args: Tuple[Any, ...], db: Optional[Database]) -> Optional[Database]:
+        if db is None:
+            for arg in args:
+                if isinstance(arg, Expr) and arg.db is not None:
+                    db = arg.db
+                    break
+        return db
 
-    """
 
-    def make_function_call(
-        *args: Expr, as_name: Optional[str] = None, db: Optional[Database] = None
-    ) -> FunctionExpr:
-        return FunctionExpr(name, args, as_name=as_name, db=db)
+_global_scope: Dict[str, _AbstractFunction] = {}
```

Comment:
Can we put `_global_scope` at the top of this file, since it's a global variable?